src/Compilers/CSharp/Portable/FlowAnalysis/VariableUsePass.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// VariableUsePass marks all referenced variables in the visited node as used (as a read).
+    /// This pass is used for things like attribute arguments and default parameter expressions,
+    /// which are otherwise not analyzed for use references.
+    /// </summary>
+    internal class VariableUsePass : BoundTreeWalkerWithStackGuard
+    {
+        /// <summary>
+        /// The current source assembly.
+        /// </summary>
+        private readonly SourceAssemblySymbol _sourceAssembly;
+
+        /// <summary>
+        /// Create a field-only visitor.
+        /// </summary>
+        /// <remarks>
+        /// Even though this class looks stateless, it cannot be shared/cached on the CSharpCompilation.
+        /// Specifically, BoundTreeWalkerWithStackGuard has an int _recursionDepth that breaks in multithreaded scenarios.
+        /// </remarks>
+        public VariableUsePass(SourceAssemblySymbol sourceAssembly)
+        {
+            _sourceAssembly = sourceAssembly;
+        }
+
+        public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+        {
+            _sourceAssembly.NoteFieldAccess(node.FieldSymbol.OriginalDefinition, true, false);
+            return base.VisitFieldAccess(node);
+        }
+    }
+
+    internal sealed class LocalVariableUsePass : VariableUsePass
+    {
+        /// <summary>
+        /// Variables that were used anywhere, in the sense required to suppress warnings about
+        /// unused variables.
+        /// </summary>
+        private readonly HashSet<LocalSymbol> _usedVariables;
+
+        /// <summary>
+        /// Variables that were used anywhere, in the sense required to suppress warnings about
+        /// unused variables.
+        /// </summary>
+        private readonly HashSet<LocalFunctionSymbol> _usedLocalFunctions;
+
+        private bool _insideNameOf;
+
+        /// <summary>
+        /// Create a visitor for use within a method (e.g. local functions).
+        /// Keep track of what local variables are referenced, by placing them in the HashSets.
+        /// Also track field references, by marking them on the SourceAssembly.
+        /// </summary>
+        public LocalVariableUsePass(SourceAssemblySymbol sourceAssembly, HashSet<LocalSymbol> usedVariables, HashSet<LocalFunctionSymbol> usedLocalFunctions)
+            : base(sourceAssembly)
+        {
+            _usedVariables = usedVariables;
+            _usedLocalFunctions = usedLocalFunctions;
+        }
+
+        public override BoundNode VisitLocal(BoundLocal node)
```

Comment:
In an earlier review, you said (in response to me implementing all such constructs):

> Please ensure we have targeted tests for all call sites of this function.

I was unable to produce an example of a case where we would use calls/etc.. You have repeatedly said that any code changes that are not tested or cannot be tested should not be included, and I was following your advice here by removing those handlers. Please clarify on how you'd like me to proceed.